core/opendaq/component/include/opendaq/component_update_context_impl.h

```diff
@@ -0,0 +1,250 @@
+    if (!this->connections.hasKey(parentId))
+    {
+        ports = Dict<IString, IBaseObject>();
+    }
+    else
+    {
+        ports = this->connections.get(parentId);
+    }
+    *connections = ports.detach();
+    return OPENDAQ_SUCCESS;
+}
+
+inline ErrCode ComponentUpdateContextImpl::removeInputPortConnection(IString* parentId)
+{
+    if (!parentId)
+        return OPENDAQ_SUCCESS;
+
+    connections->deleteItem(parentId);
+    return OPENDAQ_SUCCESS;
+}
+
+inline ErrCode ComponentUpdateContextImpl::getRootComponent(IComponent** rootComponent)
+{
+    if (!rootComponent)
+        return OPENDAQ_ERR_INVALID_ARGUMENT;
+
+    *rootComponent = this->rootComponent.addRefAndReturn();
+    return OPENDAQ_SUCCESS;
+}
+
+inline StringPtr ComponentUpdateContextImpl::getRemoteId(const std::string& globalId)
+{
+    size_t firstSlashPos = globalId.find('/');
+    if (firstSlashPos == std::string::npos) 
+    {
+        // No slash found, return the original path
+        return globalId;
+    }
+
+    // Find the position of the second slash
+    size_t secondSlashPos = globalId.find('/', firstSlashPos + 1);
+    if (secondSlashPos == std::string::npos) 
+    {
+        // Only one segment found, return an empty string
+        return "";
+    }
+
+    // Erase the first segment
+    return globalId.substr(secondSlashPos);
+}
+
+inline ErrCode ComponentUpdateContextImpl::getSignal(IString* parentId, IString* portId, ISignal** signal)
+{
+    if (parentId == nullptr)
+        return OPENDAQ_ERR_INVALID_ARGUMENT;
+    if (portId == nullptr)
+        return OPENDAQ_ERR_INVALID_ARGUMENT;
+    if (signal == nullptr)
+        return OPENDAQ_ERR_INVALID_ARGUMENT;
+
+    *signal = nullptr;
+
+    DictPtr<IString, IBaseObject> connections;
+    getInputPortConnection(parentId, &connections);
+    if (!connections.hasKey(portId))
+        return OPENDAQ_NOTFOUND;
+    
+    auto signalId = connections.get(portId);
+    auto overridenSignalId = rootComponent.getGlobalId() + getRemoteId(signalId);
+
+    Bool isCircle = false;
+    for (SizeT i = 0; i < parentDependencies.getCount(); i++)
+    {
+        const auto & parentDep = parentDependencies.getItemAt(i);
+        if (parentDep == parentId)
+        {
+            std::string deps;
+            for (SizeT j = i; j < parentDependencies.getCount(); j++)
+            {
+                deps += parentDependencies.getItemAt(j).toStdString() + " -> ";
+            }
+            auto loggerComponent = rootComponent.getContext().getLogger().getOrAddComponent("Component");
+            LOG_W("Circular dependency detected: {}{}", deps, parentDep);
+            isCircle = true;
+            break;
+        }
+    }
+
+    if (isCircle == false)
+    {
+        parentDependencies.pushBack(parentId);
+        ErrCode errCode = resolveSignalDependency(overridenSignalId, signal);
+        parentDependencies.popBack();
+        if (errCode == OPENDAQ_SUCCESS)
+            return OPENDAQ_SUCCESS;
```

Comment:
Might be worth checking for unexpected error codes, as this simply ignores every error.